### PR TITLE
feat(pipeline): pipeline engine add docker executor

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
pipeline action executor add docker executor

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJhc3NpZ25lZSI6WyIxMDAxMjA1Il19&id=301588&iterationID=1174&pId=292707&type=TASK)

#### Specified Reviewers:

/assign @your-reviewer

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |  Support pipeline docker executor for cicd            |
| 🇨🇳 中文    |     pipeline action增加docker执行器         |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: modules/pipeline/pipengine/actionexecutor/logic/job.go [hunk 1]
```diff
@@ -60,6 +61,24 @@ func ParseJobHostBindTemplate(hostPath string, clusterInfo apistructs.ClusterInf
 	return b.String(), nil
 }
 
+func GenerateDockerVolumes(job *apistructs.JobFromUser) []mount.Mount {
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/logic/job.go:64]: I think we should consider why the pkg called `logic` and if it is suitable to put docker logic here?

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+
+func init() {
+	types.MustRegister(Kind, func(name types.Name, options map[string]string) (types.ActionExecutor, error) {
+		return New(name)
+	})
+}
+
+// TODO: create docker client by executor-enviroment
+func New(name types.Name) (*DockerJob, error) {
+	cli, err := client.NewClientWithOpts(client.FromEnv)
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go:49]: what's the access mode?
add a test for usage exhibition please.

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+func (d *DockerJob) Kind() types.Kind {
+	return Kind
+}
+
+func (d *DockerJob) Name() types.Name {
+	return d.name
+}
+
+func (d *DockerJob) Status(ctx context.Context, task *spec.PipelineTask) (desc apistructs.PipelineStatusDesc, err error) {
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go:74]: add ut for this.

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+				return nil, errors.Errorf("failed to remove docker volume, name: %s", mountVol.Name)
+			}
+			logrus.Warningf("the docker job %s's volume %s in namespace %s is not found", name, mountVol.Name, job.Namespace)
+		}
+	}
+	return task.Extra.UUID, nil
+}
+
+func (d *DockerJob) Inspect(ctx context.Context, task *spec.PipelineTask) (apistructs.TaskInspect, error) {
+	return apistructs.TaskInspect{}, nil
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go:276]: anything to inspect from `docker inspect`?

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go (hunk 4 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go:276]: added, use `docker inspect`

File: modules/pipeline/pipengine/actionexecutor/logic/job.go (hunk 1 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/logic/job.go:64]: adjusted

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go (hunk 3 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go:74]: added

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go (hunk 2 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go]: remote docker and local docker

test `remote docker` by remote host, open 2375 port

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+	if err := logic.ValidateAction(task); err != nil {
+		return desc, err
+	}
+	jobName := logic.MakeJobName(task)
+	container, err := d.client.ContainerInspect(ctx, jobName)
+	if err != nil {
+		return desc, err
+	}
+	switch container.State.Status {
+	case "running", "restarting", "created":
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go:87]: does these status have const or typed status?

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+		fmt.Sprintf("DICE_MEM_LIMIT=%f", job.Memory),
+	)
+
+	// add container TerminusDefineTag env
+	if len(job.TaskContainers) > 0 {
+		env = append(env, fmt.Sprintf("%s=%s", apistructs.TerminusDefineTag, job.TaskContainers[0].ContainerID))
+	}
+
+	// active upload logs
+	env = append(env, "ACTIONAGENT_ENABLE_PUSH_LOG_TO_COLLECTOR=true")
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go:392]: ugly

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+		env = append(env, fmt.Sprintf("%s=%s", k, v))
+	}
+
+	// add docker label
+	env = append(env, "IS_DOCKER=true")
+
+	// add namespace label
+	env = append(env, fmt.Sprintf("DICE_NAMESPACE=%s", job.Namespace))
+	env = append(env,
+		fmt.Sprintf("DICE_CPU_ORIGIN=%f", job.CPU),
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go:379]: ugly. any const ?

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go (hunk 5 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go:87]: not found in docker pkg
image

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go (hunk 6 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go:407]: should move this to `executor env`, added annotation

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go (hunk 7 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go]: adjusted, const these job env params

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go (hunk 6 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go:407]: image

Comment (sfwn) [reply modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go:407]: nono, you should search globally, and found this const inside agent pkg.

**Review by sfwn — DISMISSED** [review]

File: modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go (hunk 6 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/actionexecutor/plugins/docker/docker.go]: adjusted, and modify other which use `ACTIONAGENT_ENABLE_PUSH_LOG_TO_COLLECTOR`

**Pull request merged**